crates/common/tedge_config/src/tedge_toml/tedge_config_location.rs

```diff
@@ -268,9 +277,132 @@ impl TEdgeConfigLocation {
     }
 }
 
+pub trait ConfigSources {
+    const INCLUDE_ENVIRONMENT: bool;
+}
+
+#[derive(Clone, Debug)]
+pub struct FileAndEnvironment;
+#[derive(Clone, Debug)]
+pub struct FileOnly;
+
+impl ConfigSources for FileAndEnvironment {
+    const INCLUDE_ENVIRONMENT: bool = true;
+}
+
+impl ConfigSources for FileOnly {
+    const INCLUDE_ENVIRONMENT: bool = false;
+}
+
+#[derive(Default, Debug, PartialEq, Eq)]
+#[must_use]
+pub struct UnusedValueWarnings(Vec<String>);
+
+impl UnusedValueWarnings {
+    pub fn emit(self) {
+        for warning in self.0 {
+            tracing::warn!("{warning}");
+        }
+    }
+
+    pub fn push(&mut self, warning: String) {
+        self.0.push(warning)
+    }
+}
+
+fn update_with_environment_variables(dto: &mut TEdgeConfigDto, warnings: &mut UnusedValueWarnings) {
+    for (key, value) in std::env::vars() {
+        let tedge_key = match key.strip_prefix("TEDGE_") {
+            Some("CONFIG_DIR") => continue,
+            Some("CLOUD_PROFILE") => continue,
+            Some(tedge_key) => {
+                let parsed_key = tedge_key
+                    .to_ascii_lowercase()
+                    .parse::<tedge_config::WritableKey>();
+                match parsed_key {
+                    Ok(key) => key,
+                    Err(_) => {
+                        warnings.push(format!(
+                            "Unknown configuration field {:?} from environment variable {key}",
+                            tedge_key.to_ascii_lowercase()
+                        ));
+                        continue;
+                    }
+                }
+            }
+            None => continue,
+        };
+
+        // TODO test these warnings are vaguely sensibly formatted
```

Comment:
I should have some unit tests that checks these warnings appear as intended. I believe the warnings could only ever appear in cases where we interact with profiles (which is unlikely to be seen in combination with environment variables, since env vars generally negate the need for profiles).